### PR TITLE
Features/swagger ui 2

### PR DESCRIPTION
**Pull request by Wu-Jing**
- [x] Option to turn off swagger
- [x] Make swagger honor the versioning - (We need to create multiple SwaggerHttpService with different apiDocsPath)
- [x] Add PipelineDefinitionApiMode with ActivityDefinitionApiModel

- [x] move swagger service into v10handler
- [x] root level add ui handler and put swagger site route in there

- [x] Update spec

### CODE REVIEW

File: core/src/main/scala/yumi/Job.scala [hunk 1]
```diff
@@ -13,9 +16,16 @@ case class Job(
                 endDate: Option[DateTime] = None
               )
 
+@ApiModel(description = "Resource Requirements")
```
Comment (coderdiaries) [inline core/src/main/scala/yumi/Job.scala]: Create ApiModel class for `ResourceRequirements` and move docs to server project

File: pipeline/src/main/scala/yumi/pipeline/YumiContextFactory.scala [hunk 2]
```diff
@@ -22,10 +22,14 @@ class YumiContextFactory {
 
     if(config.hasPath("yumi.spark.master")) {
       sparkConf.setMaster(config.getString("yumi.spark.master"))
+    } else {
```
Comment (coderdiaries) [inline pipeline/src/main/scala/yumi/pipeline/YumiContextFactory.scala]: Remove the else part and rely on config.

File: server/src/main/scala/yumi/server/v10/JobService.scala [hunk 3]
```diff
@@ -1,19 +1,26 @@
...
 import yumi.metastore.Metastore
 import yumi.{Job, JobStatus, ResourcesRequirements}
 
+import scala.annotation.meta.field
 import scala.concurrent.{ExecutionContext, Future}
 
+
 class JobService(metastore: Metastore, executor: Executor) {
 
+
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/v10/JobService.scala]: REmove extra line here.

File: server/src/main/scala/yumi/server/v10/SwaggerDocService.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import com.github.swagger.akka.model._
+import io.swagger.models.ExternalDocs
+
+import scala.reflect.runtime.{universe => ru}
+
+class SwaggerDocService(system: ActorSystem) extends SwaggerHttpService with HasActorSystem {
+  override implicit val actorSystem: ActorSystem = system
+  override implicit val materializer: ActorMaterializer = ActorMaterializer()
+  override val apiTypes = Seq(ru.typeOf[JobService], ru.typeOf[V10Handlers], ru.typeOf[JobHandler])
+  override val host = "localhost:8080"
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/v10/SwaggerDocService.scala]: Parametrize this through constructor

File: server/src/main/scala/yumi/server/Run.scala [hunk 5]
```diff
@@ -35,8 +37,15 @@ object Run extends YumiApp {
...
 
-    val bindingFuture = Http()
-      .bindAndHandle(handler.routes, serverHost, serverPort)
+    var showSwagger = true
+    if(config.hasPath("yumi.server.swagger")){
+      showSwagger = config.getBoolean("yumi.server.swagger")
+    }
+
+    val yumiRoutes =
+      if(showSwagger) new SwaggerDocService(system).routes ~ handler.routes else handler.routes
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/Run.scala]: reformat the code here: 
`if(...) {`
        `...`
     `} else {`
        `....`
     `}`

File: server/src/main/scala/yumi/server/Run.scala [hunk 6]
```diff
@@ -35,8 +37,15 @@ object Run extends YumiApp {
 
     val handler = new V10Handlers(system, serviceExecutionContext, jobService)
 
-    val bindingFuture = Http()
-      .bindAndHandle(handler.routes, serverHost, serverPort)
+    var showSwagger = true
+    if(config.hasPath("yumi.server.swagger")){
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/Run.scala]: No need to check if path exists here. This required option.

File: server/src/main/scala/yumi/server/v10/JobService.scala [hunk 7]
```diff
@@ -22,13 +27,14 @@ class JobService(metastore: Metastore, executor: Executor) {
     jobOption.map(toJobResponse)
   }
 
+
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/v10/JobService.scala]: Extra line.

File: server/src/main/scala/yumi/server/v10/JobService.scala [hunk 8]
```diff
@@ -22,13 +27,14 @@ class JobService(metastore: Metastore, executor: Executor) {
...
   def handle(request: ExecuteJobRequest)
             (implicit executionContext: ExecutionContext): Future[ExecuteJobResponse] = Future {
 
     val job = metastore.addJob(Job(
       id = request.name,
       name = request.name,
-      pipelineDefinition = request.pipelineDefinition.definition,
+      pipelineDefinition = write(request.pipelineDefinition),
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/v10/JobService.scala]: Add `utils: YumiUtils` constructor argument and use `utils.serialize(request.pipelineDefinition)`

File: server/src/main/scala/yumi/server/v10/JobService.scala [hunk 9]
```diff
@@ -41,7 +47,7 @@ class JobService(metastore: Metastore, executor: Executor) {
     GetJobResponse(
       id = job.id,
       name = job.name,
-      pipelineDefinition = PipelineDefinition(job.pipelineDefinition),
+      pipelineDefinition = read[ServicePipelineDefinition](job.pipelineDefinition),
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/v10/JobService.scala]: Use ` utils.deserialize[ServicePipelineDefinition](job.pipelineDefinition)` instead.

File: server/src/main/scala/yumi/server/v10/Site.scala [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package yumi.server.v10
+
+import akka.http.scaladsl.server.Directives
+
+trait Site extends Directives {
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/v10/Site.scala]: Are we using this trait?

Comment (Wu-Jing) [inline server/src/main/scala/yumi/server/v10/Site.scala]: deleted

**Comment by Wu-Jing** [comment]
@coderdiaries  Updated base on our discussion: separate ui routes, combine swagger service into v10handler, update spec, update dependency etc

**Comment by Wu-Jing** [comment]
I will take a look adding spec for ui routes

File: server/src/main/scala/yumi/server/v10/V10Handlers.scala [hunk 11]
```diff
@@ -1,14 +1,35 @@
...
-                 ) extends JobHandler {
+                   override val jobService: JobService,
+                   override implicit val materializer: ActorMaterializer,
+                   override val apiTypes: Seq[scala.reflect.runtime.universe.Type],
+                   override val host: String,
+                   override val basePath: String,
+                   override val info: Info,
+                   override val apiDocsPath: String,
+                   override val externalDocs: Some[ExternalDocs]
+                 ) extends JobHandler with SwaggerHttpService with HasActorSystem {
```
Comment (coderdiaries) [inline server/src/main/scala/yumi/server/v10/V10Handlers.scala]: Create a `Swagger` trait:
1. use it `JobHandler` trait
2. Inherits from `SwaggerHttpService` and `SwaggerHttpService`
3. Define rout field within called `swaggerRoutes`
4. move definition of `externalDocs` as a field (not constructor argument)
5. move definition of `info` as a field (not constructor argument)
6. move definition of `apiTypes` as a field (not constructor argument)
7. move definition of `apiDocsPath ` as a field (not constructor argument)

File: server/src/main/resources/application.conf [hunk 12]
```diff
@@ -6,6 +6,7 @@ yumi {
   server {
     host = "localhost"
     port = 8080
+    swagger=true
```
Comment (coderdiaries) [inline server/src/main/resources/application.conf]: always put spaces between (e.g. `swagger = true`)

File: server/src/main/scala/yumi/server/v10/V10Handlers.scala (hunk 11 above)
Comment (Wu-Jing) [inline server/src/main/scala/yumi/server/v10/V10Handlers.scala]: Done

File: server/src/main/resources/application.conf (hunk 12 above)
Comment (Wu-Jing) [inline server/src/main/resources/application.conf]: Yep, Updated

**Comment by coderdiaries** [comment]
Looks good.

**Pull request merged**